### PR TITLE
Outgoing cleanup

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR makes various cleanups to the OutgoingAsync code.

Primarily:
 - remove entirely the P = std::promise template option, since we only use std::promise
 - move makeLambdaOutgoing and makePromiseOutgoing to IceInternal, in OutgoingAsync.h
 - remove Ice/OutgoingAsync.h from Ice/Ice.h - we need it only for the implementation of proxies, so we include it in the generated .cpp files.

### CODE REVIEW

File: cpp/src/Ice/Communicator.cpp [hunk 1]
```diff
@@ -7,3 +7,26 @@
...
+
+std::future<void>
+Ice::Communicator::flushBatchRequestsAsync(CompressBatch compress)
+{
+    auto promise = ::std::make_shared<std::promise<void>>();
+    flushBatchRequestsAsync(
+        compress,
+        [promise](::std::exception_ptr ex)
+        {
+            promise->set_exception(::std::move(ex));
```
Comment (pepone) [inline cpp/src/Ice/Communicator.cpp:25]: Minor: I thought there was no point on calling std::move with exception_ptr, as per previous reviews, we should also avoid `::std` just `std`

File: cpp/src/Ice/CommunicatorI.h [hunk 2]
```diff
@@ -8,15 +8,16 @@
...
 
 #include <functional>
 
 namespace IceInternal
 {
 
 //
-// Class for handling Ice::Communicator::begin_flushBatchRequests
+// Class for handling Ice::Communicator::flushBatchRequests
+//  This class needs to be in a public header as it's used by several other classes.
```
Comment (pepone) [inline cpp/src/Ice/CommunicatorI.h:20]:
```suggestion
// This class needs to be in a public header as it's used by several other classes.
```

**Review by pepone — APPROVED** [review]
👍 Looks good!

**Review by externl — APPROVED** [review]

File: cpp/src/Ice/Communicator.cpp (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/Ice/Communicator.cpp:25]: That's correct, we should not move exception_ptr. I only "moved" this code around...

Comment (bernardnormier) [reply cpp/src/Ice/Communicator.cpp:25]: Fixed

**Pull request merged**